File: services/go-api/handler.go
package main

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"time"
)

const (
	contentType     = "Content-Type"
	applicationJson = "application/json"
)

type Handler struct {
	db *sql.DB
}

func NewHandler(db *sql.DB) *Handler {
	return &Handler{
		db: db,
	}
}

func respondJSON(w http.ResponseWriter, status int, payload any) {
	w.Header().Set(contentType, applicationJson)
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(payload)
}

func respondError(w http.ResponseWriter, status int, msg string) {
	respondJSON(w, status, map[string]string{"error": msg})
}

type RequestsResponse struct {
	Timestamp time.Time `json:"timestamp"`
	Count     int       `json:"count"`
}

type InsertRequestBody struct {
	Name string `json:"name"`
}

func (h *Handler) insertRequest(w http.ResponseWriter, r *http.Request) {
	var body InsertRequestBody
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		respondError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if body.Name != "go" && body.Name != "rust" {
		respondError(w, http.StatusBadRequest, "name must be 'go' or 'rust'")
		return
	}

	_, err := h.db.Exec(`INSERT INTO requests (api_name) VALUES ($1)`, body.Name)
	if err != nil {
		respondError(w, http.StatusInternalServerError, err.Error())
		return
	}

	w.WriteHeader(http.StatusCreated)
}

func (h *Handler) liveness(w http.ResponseWriter, r *http.Request) {
	err := h.db.Ping()
	if err != nil {
		respondError(w, http.StatusServiceUnavailable, "db ping failed")
		return
	}
	respondJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

func (h *Handler) readiness(w http.ResponseWriter, r *http.Request) {
	_, err := h.db.Exec("SELECT 1")
	if err != nil {
		respondError(w, http.StatusServiceUnavailable, "db not ready")
		return
	}
	respondJSON(w, http.StatusOK, map[string]string{"status": "ready"})
}

func (h *Handler) getRequests(w http.ResponseWriter, r *http.Request) {
	var resp RequestsResponse
	err := h.db.QueryRow(`SELECT NOW(), COUNT(*) FROM requests WHERE api_name = 'go'`).Scan(&resp.Timestamp, &resp.Count)
	if err != nil {
		respondError(w, http.StatusInternalServerError, err.Error())
		return
	}
	respondJSON(w, http.StatusOK, resp)
}

File: services/rust-api/src/handler.rs
use axum::{
    extract::State,
    http::StatusCode,
    response::Json,
};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sqlx::{FromRow, PgPool};

type JsonError = (StatusCode, Json<Value>);

fn internal_error(e: impl ToString) -> JsonError {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({"error": e.to_string()})),
    )
}

#[derive(Serialize, FromRow)]
pub struct RequestsResponse {
    pub timestamp: DateTime<Utc>,
    pub count: i64,
}

#[derive(Deserialize)]
pub struct InsertRequestBody {
    pub name: String,
}

pub async fn liveness(State(pool): State<PgPool>) -> Result<Json<Value>, JsonError> {
    sqlx::query("SELECT 1")
        .execute(&pool)
        .await
        .map_err(|_| {
            (
                StatusCode::SERVICE_UNAVAILABLE,
                Json(json!({"error": "db ping failed"})),
            )
        })?;
    Ok(Json(json!({"status": "ok"})))
}

pub async fn readiness(State(pool): State<PgPool>) -> Result<Json<Value>, JsonError> {
    sqlx::query("SELECT 1")
        .execute(&pool)
        .await
        .map_err(|_| {
            (
                StatusCode::SERVICE_UNAVAILABLE,
                Json(json!({"error": "db not ready"})),
            )
        })?;
    Ok(Json(json!({"status": "ready"})))
}

pub async fn get_requests(
    State(pool): State<PgPool>,
) -> Result<Json<RequestsResponse>, JsonError> {
    let row = sqlx::query_as::<_, RequestsResponse>(
        "SELECT NOW() as timestamp, COUNT(*) as count FROM requests WHERE api_name = 'rust'",
    )
    .fetch_one(&pool)
    .await
    .map_err(internal_error)?;

    Ok(Json(row))
}

pub async fn insert_request(
    State(pool): State<PgPool>,
    Json(body): Json<InsertRequestBody>,
) -> Result<StatusCode, JsonError> {
    if body.name != "rust" {
        return Err((
            StatusCode::BAD_REQUEST,
            Json(json!({"error": "name must be 'go' or 'rust'"})),
        ));
    }

    sqlx::query("INSERT INTO requests (api_name) VALUES ($1)")
        .bind(&body.name)
        .execute(&pool)
        .await
        .map_err(internal_error)?;

    Ok(StatusCode::CREATED)
}
